Fix lagging beatmap carousel because of storyboard-loading

Aergwyn opened the pull request:
Closes #1606 at the cost of up to ~1s longer loading screen (from ~21s to ~22s) because the storyboard parsing happens there now.

Also:
- Removed `Storyboard` from `Beatmap` and added it to `WorkingBeatmap` instead
- `BeatmapDecoder` (or `LegacyDecoder`) now splits the `Beatmap` and `Storyboard` parsing part
- loads Storyboard from .osb, or .osu if the former is not present
- expanded unit tests

Done with guidance from @FreezyLemon!

Aergwyn committed:
split storyboard loading into `GetStoryboard()`

Aergwyn committed:
remove Storyboard from Beatmap, add it to WorkingBeatmap

Aergwyn committed:
readd storyboard to beatmap + minor cleanup

Aergwyn committed:
split parsing a beatmap and parsing a storyboard

Aergwyn committed:
remove unnecessary lines

Aergwyn committed:
remove unnecessary `using`

peppy commented:
22s loading screen?! what map are you using for testing?

smoogipoo commented on osu.Game/Beatmaps/BeatmapManager.cs:
```diff
@@ -623,6 +612,26 @@ protected override Track GetTrack()
+                    BeatmapDecoder decoder;
+                    using (var stream = new StreamReader(store.GetStream(getPathForFile(BeatmapInfo.Path))))
+                        decoder = BeatmapDecoder.GetDecoder(stream);
+
```
I think it might still be desirable to parse the beatmap file with `DecodeStoryboard`, for the cases where the storyboard is not solely inside the .osb. I found the following map which does not have the storyboard in an osb: https://osu.ppy.sh/beatmapsets/797#osu/5089. Thoughts, @peppy, or should we fix these on a per-case basis?

Aergwyn replied:
I added something for that. If there is no .osb then it uses the .osu. If it's mixed it's a different thing though...

Aergwyn commented:
It's `Mili - world.execute(me);` (ID 470977 iirc) - which is really heavy on the storyboard side.

smoogipoo commented:
How about splitting out OsuLegacyDecoder? It's getting pretty chunky in its current state having to contain logic for parsing beatmaps + storyboards, and duplicates a bunch of the parsing code.

I suggest the following:
* A base LegacyDecoder class that reads lines from a beatmap and gives lines and sections.
  * A deriving LegacyBeatmapDecoder class that parses sections it needs to and provides a Beatmap output.
  * A deriving LegacyStoryboardDecoder class that parses sections it needs to and provides a Storyboard output.

Aergwyn commented:
Sounds reasonable. I'll try it once I'm home.

Aergwyn committed:
restructured OsuLegacyDecoder into LegacyDecoder

Beatmap works, Storyboard not...

Aergwyn committed:
CI adjustments

- removing unnecessary `using`s
- name Fields/Methods according to rules
- removing unnecessary initializations

Aergwyn committed:
Split retrieving of beatmap and storyboard decoder

Storyboard works again. Not satisfied with the solution though.

Aergwyn committed:
reverted split at Decoder, moved logic down

I'm done experimenting, sorry
- `Decoder` only returns a "Beatmap"`Decoder` now
- "Storyboard"`Decoder` is retrieved from a "Beatmap"`Decoder`
- moved parse methods down in the hierarchy where I forgot to do that
- renamed `OsuLegacyDecoderTest` to `LegacyDecoderTest`

Aergwyn committed:
removed unnecessary using

Aergwyn committed:
expanded BeatmapDecoder tests,  added StoryboardDecoder tests

Aergwyn committed:
fall back to .osu file for storyboard if no .osb file is present

+ CI fixes

Aergwyn committed:
fix possible NullReference?

AppVeyor pls?

peppy committed:
Merge branch 'master' into fix-beatmap-carousel-lag

peppy committed:
Fix post-merge issue
